test/EResourceTest.cpp

```diff
@@ -0,0 +1,228 @@
+        {
+            resource.release();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource acquired exclusive")
+    {
+        kf::EResource resource;
+
+        WHEN("The resource is converted to shared")
+        {
+            bool exclusiveAcquired = resource.acquireExclusive();
+            REQUIRE(exclusiveAcquired == true);
+
+            resource.convertExclusiveToShared();
+
+            THEN("The lock is converted to shared")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() > 0);
+            }
+
+            WHEN("The shared lock is released")
+            {
+                resource.release();
+
+                THEN("The resource is no longer locked")
+                {
+                    REQUIRE(resource.isAcquiredExclusive() == false);
+                    REQUIRE(resource.isAcquiredShared() == 0);
+                }
+            }
+        }
+    }
+
+    GIVEN("The EResource locked using the lock function")
+    {
+        kf::EResource resource;
+        resource.lock();
+
+        THEN("The resource is acquired exclusively and shared")
+        {
+            REQUIRE(resource.isAcquiredExclusive() == true);
+            REQUIRE(resource.isAcquiredShared() == 1);
+        }
+
+        WHEN("The resource is unlocked")
+        {
+            resource.unlock();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource is locked shared using the lock_shared function")
+    {
+        kf::EResource resource;
+        resource.lock_shared();
+
+        THEN("The resource is acquired shared")
+        {
+            REQUIRE(resource.isAcquiredExclusive() == false);
+            REQUIRE(resource.isAcquiredShared() > 0);
+        }
+
+        WHEN("The resource is unlocked shared")
+        {
+            resource.unlock_shared();
+
+            THEN("The resource is no longer locked")
+            {
+                REQUIRE(resource.isAcquiredExclusive() == false);
+                REQUIRE(resource.isAcquiredShared() == 0);
+            }
+        }
+    }
+
+    GIVEN("The EResource and 2 threads")
+    {
+        constexpr int kOneSecond = 10'000'000; // 1 second in 100-nanosecond intervals
+        struct Context
+        {
+            kf::EResource* resource;
+            bool wasAcquiredByFirst = false;
+            bool wasAcquiredBySecond = false;
+        };
+
+        WHEN("Multiple threads attempt to acquire exclusive the resource")
+        {
+            kf::EResource resource;
+            kf::Thread thread1;
```

Comment:
2 threads is not enough because the delay between acquireExclusive() and release() is extremely small. add more threads and some time gap (1ms wait) before release() call.

Reply:
Basically, the amount of what enough for the test is the maximum count of logical threads provided by CPU.